### PR TITLE
Add basic stub for characteristic

### PR DESCRIPTION
**Pull request by alexjurkiewicz**
https://characteristic.readthedocs.io/en/stable/

### CODE REVIEW

**Comment by gvanrossum** [comment]
Did you get permission from the authors/developers of the package? See https://www.python.org/dev/peps/pep-0484/#the-typeshed-repo

**Comment by hynek** [comment]
Permission granted and please look into using [attrs](http://www.attrs.org) instead. ;)

**Comment by gvanrossum** [comment]
OK, but what we have here is incomplete -- it only defines one method. @alexjurkiewicz Can you make it more complete? (Maybe run stubgen or just transcribe the major API elements by hand.

**Comment by alexjurkiewicz** [comment]
Thanks @hynek -- I apologise for my lack of legacy rework :D

I've added the remainder of the public API. How do you want the `pyi` file formatted? Some of the function definitions are extremely long, I've split them over multiple lines, but other files seem to use a single line...

**Comment by gvanrossum** [comment]
This looks fine, but could you reformat using 4-space indents? That aspect of PEP 8 still applies... :-)

**Comment by alexjurkiewicz** [comment]
whoops!

**Comment by ambv** [comment]
@hynek, the irony of the `characteristic` tagline in this context does not escape me! 😆 

@gvanrossum, doesn't the use of class decorators like `immutable` or `with_repr` erase types with the current type definition? It seems to me those class decorators should rather be annotated as:
```
T = TypeVar('T')
def with_repr(attrs: Sequence[Union[AnyStr,Attribute]]) -> Callable[[T], T]: ...
```

File: third_party/2and3/characteristic/__init__.pyi [hunk 1]
```diff
@@ -0,0 +1,34 @@
+from typing import Sequence, Callable, Union, Any, Optional, AnyStr, TypeVar, Type
+
+def with_repr(attrs: Sequence[Union[AnyStr,Attribute]]) -> Callable[..., Any]: ...
```
Comment (gvanrossum) [inline third_party/2and3/characteristic/__init__.pyi]: There should be a space after the comma between `AnyStr` and `Attribute` (here and below).

@ambv -- I'm surprised flake8 didn't catch this?

**Comment by gvanrossum** [comment]
> @gvanrossum, doesn't the use of class decorators like immutable or with_repr erase types with the current type definition? It seems to me those class decorators should rather be annotated as:
> ```python
> T = TypeVar('T')
> def with_repr(attrs: Sequence[Union[AnyStr,Attribute]]) -> Callable[[T], T]: ...
> ```

Sadly that doesn't work yet: https://github.com/python/mypy/issues/1551

**Comment by ambv** [comment]
> There should be a space after the comma between AnyStr and Attribute (here and below).
> 
> @ambv -- I'm surprised flake8 didn't catch this?
> 
Haha, this pull request predates flake8 integration. I only enabled it 3 days ago. Check it out locally and rebase, you'll see it fail the 'flake8' run.

**Comment by alexjurkiewicz** [comment]
I'll be able to work on this again in early Jan, apologies for the delay

**Comment by gvanrossum** [comment]
> Check it out locally and rebase, you'll see it fail the 'flake8' run.

I had done exactly that, and it passed. Then I found that I didn't run `pip3 install -U -r requirements-tests-py3.txt`. With that it finds these bugs.

I'm tempted to just fix the whitespace using GitHub's edit feature and merge.

**Comment by gvanrossum** [comment]
Thanks! If you want to change more, just submit a new PR.

**Comment by alexjurkiewicz** [comment]
Thanks for your help and time!

**Pull request merged**